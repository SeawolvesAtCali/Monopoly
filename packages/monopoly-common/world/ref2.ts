import { assertExist } from '@prisel/common';
import { serializable } from 'serializr';
import { GameObject } from './GameObject';
import { Id } from './Id';
import { World } from './World';

export class Ref<T extends GameObject> {
    @serializable
    id;
    world: World;

    static of<T extends GameObject>(object: T | Id<T>, world: World): Ref<T> {
        const ref = new Ref<T>(object instanceof GameObject ? object.id : object);
        ref.world = world;
        return ref;
    }

    static forTest<T extends GameObject>(object: T | Id<T>): Ref<T> {
        const ref = new Ref<T>(object instanceof GameObject ? object.id : object);
        return ref;
    }

    constructor(id: Id<T>) {
        this.id = id;
    }

    get(): T {
        return assertExist(this.world, 'world').get(this.id);
    }

    equals(ref: Ref<any>) {
        return ref.id === this.id;
    }
}
