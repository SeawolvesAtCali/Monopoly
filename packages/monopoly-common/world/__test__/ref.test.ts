import { deserialize, serialize } from 'serializr';
import { Ref } from '../ref2';

describe('ref', () => {
    test('serialize', () => {
        const ref = Ref.forTest('123');
        expect(serialize(ref)).toMatchObject({
            id: '123',
        });
    });
    test('deserialize', () => {
        const ref = Ref.forTest('123');
        const deserialized = deserialize(Ref, serialize(ref));
        expect(deserialized).toBeInstanceOf(Ref);
        expect(deserialized.equals(ref)).toBe(true);
    });
});
