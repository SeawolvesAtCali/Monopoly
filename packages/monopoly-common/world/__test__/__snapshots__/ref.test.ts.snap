// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ref serialize 1`] = `
Object {
  "data": Object {
    "ref": "object2",
  },
  "id": "object1",
  "type": "test",
}
`;
